financial-templates-lib/clients/ExpiringMultiPartyClient.js

```diff
@@ -83,37 +83,69 @@ class ExpiringMultiPartyClient {
-    this.sponsorAddresses = [...new Set(events.map(e => e.returnValues.sponsor))];
     this.cumulativeFeeMultiplier = this.toBN(cumulativeFeeMultiplier.toString());
+  }
 
-    // Fetch sponsor position, liquidation, and current time data in parallel batches, 20 at a time, to be safe and not overload the web3 node.
-    const WEB3_CALLS_BATCH_SIZE = 20;
-    const [positions, allLiquidations, currentTime] = await Promise.all([
-      Promise.map(this.sponsorAddresses, address => this.emp.methods.positions(address).call(), {
+  async update() {
+    // If it is the first run then get contract contestants. This only needs to be called once.
+    if (!this.collateralRequirement) {
+      await this.initialSetup();
+    }
+    // Fetch contract state variables in parallel.
+    const [newSponsorEvents, endedSponsorEvents, liquidationCreatedEvents, currentTime] = await Promise.all([
+      this.emp.getPastEvents("NewSponsor", { fromBlock: 0 }),
+      this.emp.getPastEvents("EndedSponsorPosition", { fromBlock: 0 }),
+      this.emp.getPastEvents("LiquidationCreated", { fromBlock: 0 }),
+      this.emp.methods.getCurrentTime().call()
+    ]);
+
+    // Array of all sponsors, over all time. Can contain repeats for every `NewSponsor` event.
+    const newSponsorAddresses = [...newSponsorEvents.map(e => e.returnValues.sponsor)];
+
+    // Array of ended sponsors. Can contain repeats for every `EndedSponsorPosition` event.
+    const endedSponsorAddresses = [...endedSponsorEvents.map(e => e.returnValues.sponsor)];
+
+    // Filter out the active sponsors by removing ended sponsors from all historic sponsors. Note that a sponsor could
+    // create a position, end their position and then create another one. They could potentially do this multiple times.
+    // As a result, create a temp array `sponsorsToRemove` which is updated as sponsors are removed from the array.
+    let sponsorsToRemove = endedSponsorAddresses; // temp array to contain address to remove.
+    this.activeSponsors = newSponsorAddresses.filter(address => {
+      // If the sponsorsToRemove contains the current address, then that address should be removed from newSponsorAddresses.
+      const shouldRemoveSponsor = !sponsorsToRemove.includes(address);
+      // Update the `sponsorsToRemove` by removing the first instance of the address from the sponsorsToRemove array.
+      const index = sponsorsToRemove.indexOf(address);
+      if (index !== -1) sponsorsToRemove = [...sponsorsToRemove.slice(0, index), ...sponsorsToRemove.slice(index + 1)];
+      return shouldRemoveSponsor;
+    });
+
+    // Array of all liquidated sponsors, over all time. Use a Set to ensure only contains unique elements.
+    const liquidatedSponsors = [...new Set(liquidationCreatedEvents.map(e => e.returnValues.sponsor))];
+
+    // Fetch sponsor position & liquidation in parallel batches, 20 at a time, to be safe and not overload the web3 node.
+    const WEB3_CALLS_BATCH_SIZE = 150;
```

Comment:
I increased this to 150. I'm pretty sure the infura nodes can take this [no problem](https://ethereum.stackexchange.com/questions/23177/how-many-concurrent-requests-can-an-ethereum-node-handle). If we bump this up we'll reduce the run time even further.

Reply:
Cool with me I have no opinion here. Perhaps @daywiss has a better idea